boards/arm/stm32/stm32f411-minimum/src/stm32_composite.c

```diff
@@ -0,0 +1,276 @@
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_USBMSC_COMPOSITE
+static void board_mscuninitialize(struct usbdevclass_driver_s *classdev)
+{
+  if (g_mschandle)
+    {
+      usbmsc_uninitialize(g_mschandle);
+    }
+
+  g_mschandle = NULL;
+}
+#endif
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: board_composite_initialize
+ *
+ * Description:
+ *   Perform architecture specific initialization of a composite USB device.
+ *
+ ****************************************************************************/
+
+int board_composite_initialize(int port)
+{
+  return OK;
+}
+
+/****************************************************************************
+ * Name:  board_composite_connect
+ *
+ * Description:
+ *   Connect the USB composite device on the specified USB device port using
+ *   the specified configuration.  The interpretation of the configid is
+ *   board specific.
+ *
+ * Input Parameters:
+ *   port     - The USB device port.
+ *   configid - The USB composite configuration
+ *
+ * Returned Value:
+ *   A non-NULL handle value is returned on success.  NULL is returned on
+ *   any failure.
+ *
+ ****************************************************************************/
+
+void *board_composite_connect(int port, int configid)
+{
+  /* Here we are composing the configuration of the usb composite device.
+   *
+   * The standard is to use one CDC/ACM and one USB mass storage device.
+   */
+
+  if (configid == 0)
+    {
+      struct composite_devdesc_s dev[2];
+      int ifnobase = 0;
+      int strbase = COMPOSITE_NSTRIDS;
+      int dev_idx = 0;
+
+#ifdef CONFIG_USBMSC_COMPOSITE
+      /* Configure the mass storage device device */
+
+      /* Ask the usbmsc driver to fill in the constants we didn't
+       * know here.
+       */
+
+      usbmsc_get_composite_devdesc(&dev[dev_idx]);
+
+      /* Overwrite and correct some values... */
+
+      /* The callback functions for the USBMSC class */
+
+      dev[dev_idx].classobject  = board_mscclassobject;
+      dev[dev_idx].uninitialize = board_mscuninitialize;
+
+      /* Interfaces */
+
+      dev[dev_idx].devinfo.ifnobase = ifnobase; /* Offset to Interface-IDs */
+      dev[dev_idx].minor = 0;                   /* The minor interface number */
+
+      /* Strings */
+
+      dev[dev_idx].devinfo.strbase = strbase;   /* Offset to String Numbers */
+
+      /* Endpoints */
+
+      dev[dev_idx].devinfo.epno[USBMSC_EP_BULKIN_IDX]  = 3;
+      dev[dev_idx].devinfo.epno[USBMSC_EP_BULKOUT_IDX] = 3;
+
+      /* Count up the base numbers */
+
+      ifnobase += dev[dev_idx].devinfo.ninterfaces;
+      strbase  += dev[dev_idx].devinfo.nstrings;
+      n++;
```

Comment:
Where `n` is declared?

Reply:
Oh no. Mass rename mistake. But I compiled it without warnings from board code. Indeed, where does it come from?
Should be dev_idx++, will fix, thanks.

Reply:
Sorry, `CONFIG_USBMSC_COMPOSITE` (and `CDCACM_COMPOSITE`) options were disabled since USBMSC experiments, hence the code for both parts was hidden behind macros. I enabled them locally, found a few more lines, renamed accordingly and compiled.
Also pushed a new defconfig for composite, looking at `stm3210e-eval:composite` and `olimexino-stm32:composite`.